core/src/main/java/org/hibernate/ogm/dialect/eventstate/impl/EventContextManager.java

```diff
@@ -72,7 +72,9 @@ void onEventFinished() {
 
 		SessionImplementor session = (SessionImplementor) states.get( SessionImplementor.class );
 
-		for ( Entry<Class<?>, Object> state : states.entrySet() ) {
+		// we need to make a copy of the map as an OperationCollector entry might be added during
```

Comment:
Yeah, I'm not a fan either. I pushed it to have your feedback.

When we use InvocationCollectingGridDialect, we add things to an `OperationCollector` which is stored in the states. When grouping, we might do it in the `OperationsQueue` onFinish call and if the `OperationCollector` is not already present in the state, we add it to the states map.